pufferlib/ocean/slimevolley/slimevolley.h

```diff
@@ -0,0 +1,615 @@
+    else {
+        agent_set_action(right, &env->actions[3]);
+    }
+
+    // Update
+    agent_update(left);
+    agent_update(right);
+
+    if (env->delay_frames == 0) {
+        ball_accelerate(ball, 0, GRAVITY);
+        ball_limit_speed(ball, 0, MAX_BALL_SPEED);
+        ball_move(ball);
+    }
+    else {
+        env->delay_frames--;
+    }
+
+    if (ball_is_colliding(ball, (SphericalObject*)left)){
+        ball_bounce(ball, (SphericalObject*)left);
+    }
+    if (ball_is_colliding(ball, (SphericalObject*)right)){
+        ball_bounce(ball, (SphericalObject*)right);
+    }
+    if (ball_is_colliding(ball, (SphericalObject*)env->fence_stub)){
+        ball_bounce(ball, (SphericalObject*)env->fence_stub);
+    }
+
+    int right_reward = -ball_check_edges(ball);
+
+    if (right_reward != 0){
+        new_match(env);
+        if (right_reward == -1){
+            right->lives--;
+            env->rewards[0] = 1.0f;
+            if (env->num_agents == 2){
+                env->rewards[1] = -1.0f;
+            }
+        }
+        else{
+            left->lives--;
+            env->rewards[0] = -1.0f;
+            if (env->num_agents == 2){
+                env->rewards[1] = 1.0f;
+            }
+        }
+    }
+    agent_update_state(left, ball, right);
+    agent_update_state(right, ball, left);
+
+    if (env->tick > MAX_TICKS || left->lives <= 0 || right->lives <= 0){
+        env->terminals[0] = 1;
+        if (env->num_agents == 2){
+            env->terminals[1] = 1;
+        }
+        env->log.perf = (left->lives - right->lives + 5.0f)  / 10.0f; // normalize to 0-1
+        env->log.score = (float)(left->lives - right->lives);
+        env->log.episode_return = (5.0f - right->lives);
+        env->log.episode_length = (float)env->tick;
+        env->log.n++;
+        c_reset(env);
+    }
+    
+}
+
+// Required function. Should handle creating the client on first call
+void c_render(SlimeVolley* env) {
+    if (!IsWindowReady()) {
+        InitWindow(WINDOW_WIDTH, WINDOW_HEIGHT, "PufferLib SlimeVolley");
+        SetTargetFPS(60);
+    }
+
+    // Standard across our envs so exiting is always the same
+    if (IsKeyDown(KEY_ESCAPE)) {
+        exit(0);
+    }
+    BeginDrawing();
+    ClearBackground(BACKGROUND_COLOR);
+    wall_display(env->ground);
+    wall_display(env->fence);
+    ball_display(env->fence_stub);
+    ball_display(env->ball);
+
+    for (int i=0; i<2; i++) {
+        agent_display(&env->agents[i], env->ball->x, env->ball->y);
+    }
+
+    EndDrawing();
+}
+
+// Required function. Should clean up anything you allocated
+// Do not free env->observations, actions, rewards, terminals
+void c_close(SlimeVolley* env) {
+    free(env->agents);
+    free(env->ground);
+    free(env->fence);
+    free(env->fence_stub);
+    free(env->ball);
+    if (env->num_agents == 1) {
+        free(env->bot_observations);
+        free(env->bot_actions);
+    }
+    if (IsWindowReady()) {
+        CloseWindow();
+    }
+}
```

Comment:
Pretty good first env PR. Mergeable. Biggest thing is overmodularization. We're neurotic about keeping env code as simple as possible. Top to bottom logic with minimal extra functions where required to avoid duplicating big chunks of code. Reason is that the penalty for complexity is 10x higher in RL. Bugs don't come out in tests. They come out in wacky learning curves etc. So where level of abstraction is a bit more subjective in general software within somewhat wider bounds, it's much more towards less abstraction for RL sim code